### PR TITLE
[GEP-21] logging extension

### PR DESCRIPTION
**Pull request by nickytd**
/area logging
/kind discussion

**What this PR does / why we need it**:
This PR opens discussion for providing logging stack in gardener as an extension

### CODE REVIEW

File: docs/proposals/21-logging-extension.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+Although intended in the future, following goals are not target of the current proposal:
+
+- migrating logging stack components to other alternatives
+- introducing logging stack components operators such as loki-operator or fluent-bit operator
+
+## Proposal
+
+![logging extension](assets/21-logging-extension.png)
+
+### Logging CRD
```
Comment (rfranzke) [inline docs/proposals/21-logging-extension.md:55]: I don't like this approach too much. Extracting the logging components does not make too much sense since it's very unlikely that there will be more than one implementation for this API. Also, it doesn't really solve any problems but just creates more code, boilerplate, and complexity. The background is very similar to why GEP-19 is what it is (monitoring is not moved out into an extension).

In the end, g/g still needs specific knowledge about which logging stack is installed since it has to provide the parsers and filters for the components it deploys in the correct format (e.g. https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/component/kubeapiserver/logging.go).

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: docs/proposals/21-logging-extension.md (hunk 1 above)
Comment (nickytd) [reply docs/proposals/21-logging-extension.md:55]: The only managed and predictable way to switch logging stacks in the near future is to move out the entire lifecycle following the extension architecture. Current implementation does not allow that. Further more migrating towards logging stack operators is the next step as at the moment this tight application specific knowledge is coded in the core. Take a fluent-bit as an example. At the moment there are static filter, parser definitions in the code. The example above is a perfect illustration of this. The operator abstracts those definitions into resources and provides aggregation and validation functions. The same holds for Grafana or for any other application specific component which is brought externally. 
So extensions are the correct way to externalise the application specific lifecycle from core. That holds valid for prometheus as well. The monitoring shall be moved out as well, introducing the prometheus operator APIs. The latter is a fundamental building block if components want to have standard way to supply scrape configs, rules, alerts and so on.
Operators for external components bringing corresponding APIs is the way forward. The trick question is how to introduce those in predictable, manageable and testable way and here the gardener extensions mechanism is a perfect fit.

Comment (Kristian-ZH) [reply docs/proposals/21-logging-extension.md:55]: Currently, all extensions need to know about the logging stack because they bring their own parsers and filters. We just add the same knowledge to the Gardener.
All other things like installing resources in the seed cluster and systemd services in the shoot are things that other extensions are already doing. Also as @nickytd said, we think that the logging stack should be easy pluggable and changeable part of the puzzle because we see in the future even scenarios like the user  to be able to choose which logging stack he wants, and the extension will deploy it.

Also, we plan different logging upgrades in different paths. This means that a lot of new code will be created/replaced in the repo and this is also why we wanted to have our own lifecycle

Comment (rfranzke) [reply docs/proposals/21-logging-extension.md:55]: > The only managed and predictable way to switch logging stacks in the near future is to move out the entire lifecycle following the extension architecture. Current implementation does not allow that.

Why not? It doesn't matter much where the code resides (this repository or another one).

> at the moment this tight application specific knowledge is coded in the core. Take a fluent-bit as an example. At the moment there are static filter, parser definitions in the code. The example above is a perfect illustration of this.

What do you mean here? I don't see any abstraction proposal in the document. But yes, g/g still has to deploy the components - and with them the logging (and monitoring) configuration. Those configurations are specific to the concrete logging (fluent-bit filters/parsers) or monitoring (Prometheus/Alertmanger scrape configs/alerting rules) stacks. Unless you come up with an abstraction for g/g that fits for all possible logging/monitoring implementations, moving out the stacks is not sustainable from a development PoV. Even if you introduce such abstractions, it's probably becoming more complex in the end.

> at holds valid for prometheus as well. The monitoring shall be moved out as well, introducing the prometheus operator APIs. The latter is a fundamental building block if components want to have standard way to supply scrape configs, rules, alerts and so on.
> Operators for external components bringing corresponding APIs is the way forward.

We decided against exactly this as part of GEP-19. Deploying the prometheus-operator externally while g/g still has to create the respective CRs for this specific monitoring implementation does not make sense and just creates more complexity (see above).

> Currently, all extensions need to know about the logging stack because they bring their own parsers and filters. We just add the same knowledge to the Gardener.

Same like above - unless you bring abstractions for both g/g and the extensions, you don't improve things but just introduce more complexity. With extensions, it would be even worse, because they might then have dependencies on both g/g and the logging extension.

> Also as @nickytd said, we think that the logging stack should be easy pluggable and changeable part of the puzzle because we see in the future even scenarios like the user to be able to choose which logging stack he wants, and the extension will deploy it.

Why should this ever be a user decision? The logging stack runs solely in the seed and we don't provide application logging for the user workload, so it should not be in the interest of an end-user to change the stack or its configuration.

> Also, we plan different logging upgrades in different paths. This means that a lot of new code will be created/replaced in the repo and this is also why we wanted to have our own lifecycle

Don't get me wrong, I am the first one who would love to see the monitoring/logging specifics being out of g/g. However, I can only support it if we follow a sustainable approach here that eliminates today's problems without introducing new dependencies or issues. Maybe I didn't understand all your plans yet, but the proposal as such is not too detailed yet.

**Comment by nickytd** [comment]
At the moment,  monitoring and logging stacks orchestration is tightly coupled with gardener. It is not open neither for modifications of those 2 stack nor for easy modifications and independent release cycles. There are coupled and hidden dependencies.

Proven by time and good principles, any k8s workload these days has a dedicated provisioning component which usually express requirements towards other subsystems via APIs 
And here are some examples:
- Prometheus monitoring stacks usually provide monitoring APIs: [monitoring.coreos.com](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com%2fv1)

- logging stacks featuring fluent-bit log shippers provide APIs such as: https://github.com/fluent/fluent-operator/tree/master/apis/fluentbit/v1alpha2

Looking into some other workloads that depend on those APIs.
such as nginx ingress controller we will spot constructs [like](https://github.com/bitnami/charts/blob/4b447271e7e9ee2658dbca6a4415b800b1dc73e7/bitnami/nginx-ingress-controller/values.yaml#L922)
```yaml
  ## Prometheus Operator ServiceMonitor configuration
  ##
  serviceMonitor:
    ## @param metrics.serviceMonitor.enabled Create ServiceMonitor resource for scraping metrics using PrometheusOperator
    ##
    enabled: false
    ## @param metrics.serviceMonitor.namespace Namespace in which Prometheus is running
```
In this way the ingress controller can express its dependency on the monitoring stack via the monitoring api.

Going back n the gardener context, gardener core and other extensions can express the same types of dependencies based on clear APIs instead of tight coupling. Orchestration of such subsystems belongs to their dedicated operators which have the best knowledge how to manage them. 

So in short there are three important aspects of this proposal:
* solve the tight coupling in the provisioning of logging stack
* allow flexibility in designs of the logging stack especially if you need different layouts based on specific clusters sizes workloads, HA or Disaster Recovery requirements
* declare dependencies among parties based on APIs

Or just another example supporting those principles: workloads that require certificates and trusted TLS now express dependencies towards [cert-manager.io APIs](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io%2fv1 )

To support this case and to validate the implementation there is a running PoC. We will publish links here in the next days with the repo and schedule a meeting to discuss our findings.

File: docs/proposals/21-logging-extension.md (hunk 1 above)
Comment (nickytd) [reply docs/proposals/21-logging-extension.md:55]: Yes we have a running PoC and we can go deep dive into it.

**Comment by rfranzke** [comment]
> Or logging stacks featuring fluent-bit log shippers provide APIs such as: https://github.com/fluent/fluent-operator/tree/master/apis/fluentbit/v1alpha2

Nice, interesting, that looks similar to what GEP-19 tries to achieve.

> So if we go back in the gardener context, gardener core and other extensions can express the same types of dependencies based on clear APIs instead of tight coupling. Orchestration of such functions belongs to their dedicated operators that has the best knowledge how to manage them.

My problem is that neither the logging nor monitoring stack can become "real" extensions this way. Today, for Gardener the extension APIs can be implemented by any concrete type Gardener itself does not know or care about (e.g., `Network` extension has `calico` or `cilium` implementations). The core code base does not have any assumptions whatsoever what the concrete type is (there are no Calico/Cilium-specifics in g/g).

If we now introduced such `Logging` or `Monitoring` CRDs as suggested in this proposal, g/g would still have hard dependencies to Prometheus or fluent-bit, hence to the concrete types/implementations. You wouldn't be able to install a different logging/monitoring extension type - it would always need to be Prometheus/fluent-bit. Hence, there would not be real extensibility here - it would rather only be the externalization of the deployment of the stacks. This however does increase development and operational complexity from my point of view and does not solve any problems but just creates new ones.

> To support this case and to validate the implementation there is a running PoC. We will publish links here in the next days with the repo and schedule a meeting to discuss our findings.

It is obvious that this proposal can be implemented (no PoC needed), the question rather is whether it should be done this way (which I doubt).

**Comment by Kristian-ZH** [comment]
Then what about if we move the filters and the parsers which are currently in g/g to be part of the extension?
Then there will no be any dependencies in g/g about which logging stack to be used.  Our intention is to use the ProviderConfig specification in the Logging resource to tell the extension which logging stack should be deployed (OFC this will be the future cuz we will support only one option now) but this is the way how we imagine that the things should happen.
I think that this is the way how the current extension work. Gardener deploys the CRD, then deploys the Resource and the extension reconciles this resource and uses the actuator which will deploy the correct components.

The problem to move these filters and parsers to the extension is that the chance for them to become outdated is really huge and that's the reason why every extension has its own filters and parsers.
But if you think that the log format is not something that will be changed often I think that it also makes sense to keep them in the extension repo.

**Comment by nickytd** [comment]
The problem that it solves is that allows step based approach towards operators of those stack. 
1. Externalise the orchestration (as correctly described above) without introducing additional APIs vocabulary. Simplify current core orchestration implementation
2. Introduce APIs and operators in already externalised components
3. Develop flexible deployment layouts depending on cluster sizes and requirements (HA, DR, multizone setups and so on)

**Comment by rfranzke** [comment]
@Kristian-ZH This will likewise increase complexity. g/g still deploys some components, but the logging configuration results in another repository, and while deploying you have to make sure that both versions (g/g and logging extension) fit together. That doesn't sound like a reasonable setup.

> Our intention is to use the ProviderConfig specification in the Logging resource to tell the extension which logging stack should be deployed (OFC this will be the future cuz we will support only one option now) but this is the way how we imagine that the things should happen.
> Gardener deploys the CRD, then deploys the Resource and the extension reconciles this resource and uses the actuator which will deploy the correct components.

This also doesn't quite fit to the other existing extension APIs - here, the type is specifying which extension is installed. For example, when you want to switch from calico to cilium then you don't change the `.spec.networking.providerConfig` but you change the `.spec.networking.type`.
This approach only works when there is only one extension which can support many different concrete types (which is also an anti-pattern and contradicts to the motivation why we have done extensibility back then in the first place)).

@nickytd I don't see why this should be a prerequisite. You can achieve the same by following the same approach as for the monitoring stack (ref GEP-19). By the way, we basically discussed the very same questions already in #6151. Did you talk to @istvanballok and @wyb1 already?

**Comment by nickytd** [comment]
Yes that's why the review list on top is such. End results might be achieved in many ways and here we are proposing: step based manageable approach. I don't want to open this question here but imagine you want to put thanos side cars, compactors with mem-cache and corresponding ingresses next to the prometheus. This is just is another risk and increased complexity if not externalised. Or imagine the cluster sizes are so big we need to introduce message brokers to facilitate streams towards slow performing external systems such as audit logs services do you really want to keep this complexity in or rather be externalised. Yes I agree the decisions we make for logging stack most probably are relevant for monitoring stack as well.

**Comment by nickytd** [comment]
A new proposal is in progress.
